// zekrom/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.activeWin = exports.utils = exports.runScript = exports.robotKeys = exports.clipboard = exports.applescript = exports.isLinux = exports.isMac = exports.isWindows = exports.platformName = exports.Robot = exports.Iohook = exports.keycode = undefined;

var _keycode = require('./keycode');

var _keycode2 = _interopRequireDefault(_keycode);

var _iohook = require('./iohook');

var _iohook2 = _interopRequireDefault(_iohook);

var _robot = require('./robot');

var _robot2 = _interopRequireDefault(_robot);

var _applescript = require('./applescript');

var _applescript2 = _interopRequireDefault(_applescript);

var _clipboard = require('./clipboard');

var _clipboard2 = _interopRequireDefault(_clipboard);

var _robotKeys = require('./robotKeys');

var _robotKeys2 = _interopRequireDefault(_robotKeys);

var _runScript = require('./runScript');

var _runScript2 = _interopRequireDefault(_runScript);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

var _activeWin = require('./activeWin');

var _activeWin2 = _interopRequireDefault(_activeWin);

var _platform = require('./platform');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.keycode = _keycode2.default;
exports.Iohook = _iohook2.default;
exports.Robot = _robot2.default;
exports.platformName = _platform.platformName;
exports.isWindows = _platform.isWindows;
exports.isMac = _platform.isMac;
exports.isLinux = _platform.isLinux;
exports.applescript = _applescript2.default;
exports.clipboard = _clipboard2.default;
exports.robotKeys = _robotKeys2.default;
exports.runScript = _runScript2.default;
exports.utils = _utils2.default;
exports.activeWin = _activeWin2.default;
