// zekrom/lib/robotKeys.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = {
  modifier: {
    COMMAND: 'command',
    ALT: 'alt',
    CONTROL: 'control',
    SHIFT: 'shift'
  },
  default: {
    Q: 'q',
    W: 'w',
    E: 'e',
    R: 'r',
    T: 't',
    Y: 'y',
    U: 'u',
    I: 'i',
    O: 'o',
    P: 'p',
    A: 'a',
    S: 's',
    D: 'd',
    F: 'f',
    G: 'g',
    H: 'h',
    J: 'j',
    K: 'k',
    L: 'l',
    Z: 'z',
    X: 'x',
    C: 'c',
    V: 'v',
    B: 'b',
    N: 'n',
    M: 'm',
    1: '1',
    2: '2',
    3: '3',
    4: '4',
    5: '5',
    6: '6',
    7: '7',
    8: '8',
    9: '9',
    0: '0',
    BACKSPACE: 'backspace',
    DELETE: 'delete',
    ENTER: 'enter',
    TAB: 'tab',
    ESCAPE: 'escape',
    UP: 'up',
    DOWN: 'down',
    RIGHT: 'right',
    LEFT: 'left',
    HOME: 'home',
    END: 'end',
    PAGEUP: 'pageup',
    PAGEDOWN: 'pagedown',
    F1: 'f1',
    F2: 'f2',
    F3: 'f3',
    F4: 'f4',
    F5: 'f5',
    F6: 'f6',
    F7: 'f7',
    F8: 'f8',
    F9: 'f9',
    F10: 'f10',
    F11: 'f11',
    F12: 'f12',
    COMMAND: 'command',
    ALT: 'alt',
    CONTROL: 'control',
    SHIFT: 'shift',
    RIGHT_SHIFT: 'right_shift',
    SPACE: 'space',
    PRINTSCREEN: 'printscreen',
    INSERT: 'insert',
    AUDIO_MUTE: 'audio_mute',
    AUDIO_VOL_DOWN: 'audio_vol_down',
    AUDIO_VOL_UP: 'audio_vol_up',
    AUDIO_PLAY: 'audio_play',
    AUDIO_STOP: 'audio_stop',
    AUDIO_PAUSE: 'audio_pause',
    AUDIO_PREV: 'audio_prev',
    AUDIO_NEXT: 'audio_next',
    // Linux only
    AUDIO_REWIND: 'audio_rewind',
    AUDIO_FORWARD: 'audio_forward',
    AUDIO_REPEAT: 'audio_repeat',
    AUDIO_RANDOM: 'audio_random',
    // No Linux support
    NUMPAD_0: 'numpad_0',
    NUMPAD_1: 'numpad_1',
    NUMPAD_2: 'numpad_2',
    NUMPAD_3: 'numpad_3',
    NUMPAD_4: 'numpad_4',
    NUMPAD_5: 'numpad_5',
    NUMPAD_6: 'numpad_6',
    NUMPAD_7: 'numpad_7',
    NUMPAD_8: 'numpad_8',
    NUMPAD_9: 'numpad_9',
    // No Windows support
    LIGHTS_MON_UP: 'lights_mon_up',
    LIGHTS_MON_DOWN: 'lights_mon_down',
    LIGHTS_KBD_TOGGLE: 'lights_kbd_toggle',
    LIGHTS_KBD_UP: 'lights_kbd_up',
    LIGHTS_KBD_DOWN: 'lights_kbd_down'
  }
};
